### PR TITLE
[CELEBORN-1489] Update Flink support with authentication support

### PR DESCRIPTION
**Pull request by mridulm**
### What changes were proposed in this pull request?
Fix authentication support for Apache Flink.

### Why are the changes needed?
Without these changes, Apache Flink applications fail when Celeborn cluster has authentication enabled.

### Does this PR introduce _any_ user-facing change?

Fixes authentication support for Apache Flink integration

### How was this patch tested?

This is forward port + adaptation of changes we did internally (against 0.4) when testing Apache Flink applications against Celeborn cluster with authentication (and TLS) enabled.
Integration test has been updated to additionally test for Flink applications with authentication enabled in Celeborn cluster.

### CODE REVIEW

**Comment by mridulm** [comment]
+CC @otterc, @SteNicholas, @RexXiong

**Comment by mridulm** [comment]
+CC @venkata91

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/network/FlinkTransportClientFactory.java [hunk 1]
```diff
@@ -53,7 +56,7 @@ public TransportClient createClientWithRetry(String remoteHost, int remotePort)
     while (retryCount > 0) {
       try {
         return createClient(remoteHost, remotePort);
-      } catch (IOException e) {
+      } catch (Exception e) {
```
Comment (mridulm) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/network/FlinkTransportClientFactory.java:59]: SASL failures are not checked exceptions - we have to retry for those too.

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java [hunk 2]
```diff
@@ -565,6 +587,15 @@ public void setDataClientFactory(TransportClientFactory dataClientFactory) {
   @Override
   @VisibleForTesting
   public TransportClientFactory getDataClientFactory() {
+    initializeTransportClientFactoryIfRequired();
```
Comment (mridulm) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java:590]: `getDataClientFactory` gets invoked without `setupLifecycleManagerRef` primarily for tests.

File: pom.xml [hunk 3]
```diff
@@ -36,7 +36,7 @@
     <module>service</module>
     <module>master</module>
     <module>worker</module>
-    <module>web</module>
+    <!-- <module>web</module> -->
```
Comment (mridulm) [inline pom.xml:39]: Will revert this

File: common/src/test/java/org/apache/celeborn/common/network/TestHelper.java [hunk 4]
```diff
@@ -30,4 +36,19 @@ public static CelebornConf updateCelebornConfWithMap(CelebornConf conf, Map<Stri
...
+      URL url = SslSampleConfigs.class.getResource(path);
+      if (null == url) {
+        throw new IllegalArgumentException("Unable to find " + path);
+      }
+      FileUtils.copyInputStreamToFile(url.openStream(), tempFile);
+      return tempFile.getCanonicalPath();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to resolve path " + path, e);
+    }
+  }
```
Comment (mridulm) [inline common/src/test/java/org/apache/celeborn/common/network/TestHelper.java:53]: This is strictly not required for this PR - but keeps failing in `ApiBaseResourceAuthenticationSuite` (with maven) without it.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/WorkerSource.scala [hunk 5]
```diff
@@ -95,8 +95,10 @@ class WorkerSource(conf: CelebornConf) extends AbstractSource(conf, MetricsSyste
   def connectionInactive(client: TransportClient): Unit = {
     val applicationIds = appActiveConnections.remove(client.getChannel.id().asLongText())
     incCounter(ACTIVE_CONNECTION_COUNT, -1)
-    applicationIds.asScala.foreach(applicationId =>
-      incCounter(ACTIVE_CONNECTION_COUNT, -1, Map(applicationLabel -> applicationId)))
+    if (null != applicationIds) {
```
Comment (mridulm) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/WorkerSource.scala:98]: This was causing NPE in the tests

**Comment by venkata91** [comment]
> +CC @venkata91

Thanks for tagging me here. Will review the PR shortly.

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java (hunk 2 above)
Comment (venkata91) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java:590]: nit: `initTransportClientFactoryIfRequired`?

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java [hunk 6]
```diff
@@ -133,15 +136,33 @@ public FlinkShuffleClientImpl(
...
+    this.context =
         new TransportContext(
             dataTransportConf, readClientHandler, conf.clientCloseIdleConnections());
-    this.flinkTransportClientFactory =
-        new FlinkTransportClientFactory(context, conf.clientFetchMaxRetriesForEachReplica());
     this.setupLifecycleManagerRef(driverHost, port);
     this.driverTimestamp = driverTimestamp;
   }
 
+  private void initializeFlinkTransportClientFactory() {
```
[the remark this answers is not in the document]
Comment (venkata91) [reply client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java:146]: Basically similar to `initDataClientFactoryIfNeeded`?

**Review by venkata91 — CHANGES_REQUESTED** [review]
Overall lgtm except for some minor comments. This looks great, thanks for adding SASL auth support for Flink with Celeborn.

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (RexXiong) [reply client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java:146]: +1

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/network/FlinkTransportClientFactory.java [hunk 7]
```diff
@@ -63,7 +66,13 @@ public TransportClient createClientWithRetry(String remoteHost, int remotePort)
...
             e);
         if (retryCount == 0) {
-          throw e;
+          if (e instanceof InterruptedException) {
+            throw (InterruptedException) e;
+          } else if (e instanceof IOException) {
+            throw (IOException) e;
+          } else {
+            Utils.rethrowAsRuntimeException(e);
+          }
```
[the remark this answers is not in the document]
Comment (mridulm) [reply client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/network/FlinkTransportClientFactory.java:75]: Interesting java 7 feature - I was only aware of the `|` in `catch` - not that same thing applied to more precise rethrows via `if` as well !

Will change to this.

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java [hunk 8]
```diff
@@ -565,6 +587,15 @@ public void setDataClientFactory(TransportClientFactory dataClientFactory) {
...
   @VisibleForTesting
   public TransportClientFactory getDataClientFactory() {
+    initializeTransportClientFactoryIfRequired();
     return flinkTransportClientFactory;
   }
+
+  private void initializeTransportClientFactoryIfRequired() {
+    if (null == flinkTransportClientFactory) {
+      logger.info(
+          "FlinkTransportClientFactory has not been initialized, " + "initializing with default");
```
[the remark this answers is not in the document]
Comment (mridulm) [reply client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java:597]: I had it in two lines - but looks like spotless combined two lines into one - but left the concat :-)

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mridulm) [reply client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java:146]: I went back and forth on this a bit - I had started with `flinkTransportClientFactory` as an `AtomicReference` for threadsafety and ensure visibility across threads ...
But looking at `ShuffleClientImpl`, this looked like unnecessary complication.

`initializeTransportClientFactoryIfRequired` vs `initializeFlinkTransportClientFactory` is an artifact of this ... which looks like I forgot to clean up.

Thanks for catching !

**Comment by mridulm** [comment]
Thanks @venkata91, @RexXiong ... have addressed the comments.

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mridulm) [reply client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/readclient/FlinkShuffleClientImpl.java]: Thinking more, this log message is not very useful - I am removing it.

**Review by venkata91 — APPROVED** [review]

**Review by RexXiong — APPROVED** [review]
Thanks, LGTM

**Comment by mridulm** [comment]
Thanks @venkata91, @RexXiong !

**Comment by mridulm** [comment]
Let me backport this to 0.5 branch as well @RexXiong

**Pull request closed without merging**